### PR TITLE
Convert System.Runtime.Tests to use RemoteInvoke when changing CurrentCulture

### PR DESCRIPTION
**Pull request by sepidehMS**
Fixes the flakiness in System.Runtime.Tests in UWP F5. That was caused by some tests changing CurrentCulture in the same process, so it could observe wrong value at times.

cc: @tarekgh

### CODE REVIEW

**Comment by stephentoub** [comment]
I'm not following.  Why is it necessary to do this out of process?  Changing the current culture should be specific to that thread; as long as the changes are undone as part of the test (and it looks like the tests are correctly doing that), there shouldn't be any need to go out of proc.  What am I missing?

**Comment by sepidehMS** [comment]
@stephentoub My understanding was that some race condition is going on. In some runs CurrentCulture's value does not match the culture the test itself assigned it to just before running culture-sensitive code and would cause unexpected behavior.

**Comment by tarekgh** [comment]
> Changing the current culture should be specific to that thread

This is not true for UWP, in UWP we used WinRT to set the current culture there which is kind of global to the all app as WinRT doesn't have the concept of thread cultures at all.

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 1]
```diff
@@ -1017,32 +1018,49 @@ public static IEnumerable<object[]> ToString_TestData()
...
             {
-                bool isDefaultProvider = provider == null;
-                if (string.IsNullOrEmpty(format) || format.ToUpperInvariant() == "G")
+                foreach (var testdata in ToString_TestData())
                 {
-                    if (isDefaultProvider)
+                    CultureInfo originalCulture = CultureInfo.CurrentCulture;
+                    try
+                    {
+                        CultureInfo.CurrentCulture = CultureInfo.InvariantCulture;
```
Comment (tarekgh) [inline src/System.Runtime/tests/System/DecimalTests.cs]: > CultureInfo.CurrentCulture = CultureInfo.InvariantCulture; [](start = 23, length = 59)

why we are doing this inside the loop and then reset it again in the finally block?

File: src/System.Runtime/tests/System/DoubleTests.cs [hunk 2]
```diff
@@ -236,32 +237,50 @@ public static IEnumerable<object[]> ToString_TestData()
...
-                if (string.IsNullOrEmpty(format) || format.ToUpperInvariant() == "G")
+                foreach (var testdata in ToString_TestData())
                 {
-                    if (isDefaultProvider)
+                    CultureInfo originalCulture = CultureInfo.CurrentCulture;
+                    try
                     {
-                        Assert.Equal(expected, d.ToString());
-                        Assert.Equal(expected, d.ToString((IFormatProvider)null));
+                        CultureInfo.CurrentCulture = CultureInfo.InvariantCulture;
```
Comment (tarekgh) [inline src/System.Runtime/tests/System/DoubleTests.cs]: >CultureInfo.CurrentCulture = CultureInfo.InvariantCulture; [](start = 24, length = 58)

same as previous comment

File: src/System.Runtime/tests/System/StringTests.cs [hunk 3]
```diff
@@ -1293,147 +1293,295 @@ public static void IndexOf_AllSubstrings(string s, string value, int startIndex,
...
+                    CultureInfo.CurrentCulture = new CultureInfo("tr-TR");
+
+                    string value = "\u0130";
+                    Assert.Equal(19, s.IndexOf(value));
+                    Assert.Equal(19, s.IndexOf(value, StringComparison.CurrentCulture));
+                    Assert.Equal(4, s.IndexOf(value, StringComparison.CurrentCultureIgnoreCase));
+                    Assert.Equal(19, s.IndexOf(value, StringComparison.Ordinal));
+                    Assert.Equal(19, s.IndexOf(value, StringComparison.OrdinalIgnoreCase));
+
+                    value = "\u0131";
```
Comment (tarekgh) [inline src/System.Runtime/tests/System/StringTests.cs]: I believe executing all code of this method inside one RemoteInvoke will just work. also you don't have to reset the current culture back after every part of the test

File: src/System.Runtime/tests/System/StringTests.cs [hunk 4]
```diff
@@ -1293,147 +1293,295 @@ public static void IndexOf_AllSubstrings(string s, string value, int startIndex,
...
+
+            RemoteInvoke((source, target) =>
             {
-                Assert.Equal(-1, source.IndexOf(target));
-                Assert.Equal(-1, source.IndexOf(target, StringComparison.CurrentCulture));
-                Assert.Equal(-1, source.IndexOf(target, StringComparison.CurrentCultureIgnoreCase));
-            });
+                CultureInfo originalCulture = CultureInfo.CurrentCulture;
+                try
+                {
```
Comment (tarekgh) [inline src/System.Runtime/tests/System/StringTests.cs]: same as previous comment

File: src/System.Runtime/tests/System/DecimalTests.cs (hunk 1 above)
Comment (sepidehMS) [inline src/System.Runtime/tests/System/DecimalTests.cs]: That's a valid point. We can totally do that outside the loop. Will change it.

Comment (tarekgh) [inline src/System.Runtime/tests/System/DecimalTests.cs]: Please do that in all cases in your change which has the same pattern.

**Comment by stephentoub** [comment]
> This is not true for UWP, in UWP we use WinRT to set the current culture there which is kind of global to the app as WinRT doesn't have the concept of thread cultures at all.

Ouch.  That directly violates even what's documented about APIs like CultureInfo.CurrentCulture.

**Comment by tarekgh** [comment]
> Ouch. That directly violates even what's documented about APIs like CultureInfo.CurrentCulture.

Not really. although the documentation is not updated but it mention the behavior of the current culture when running in app containers

https://msdn.microsoft.com/en-us/library/system.globalization.cultureinfo.currentculture(v=vs.110).aspx#WindowsAndCurr

**Comment by sepidehMS** [comment]
@tarekgh Applied all changes related to your comments. Also removed setting CurrentCulture back to original value as it doesn't seem to be necessary.

**Comment by tarekgh** [comment]
by the way, I have logged the issue https://github.com/dotnet/docs/issues/2488 to update the docs

File: src/System.Runtime/tests/System/StringTests.cs [hunk 5]
```diff
@@ -1293,147 +1293,157 @@ public static void IndexOf_AllSubstrings(string s, string value, int startIndex,
...
-            string s = "Turkish I \u0131s TROUBL\u0130NG!";
-            Helpers.PerformActionWithCulture(new CultureInfo("tr-TR"), () =>
-            {
-                string value = "\u0130";
-                Assert.Equal(19, s.IndexOf(value));
-                Assert.Equal(19, s.IndexOf(value, StringComparison.CurrentCulture));
-                Assert.Equal(4, s.IndexOf(value, StringComparison.CurrentCultureIgnoreCase));
-                Assert.Equal(19, s.IndexOf(value, StringComparison.Ordinal));
-                Assert.Equal(19, s.IndexOf(value, StringComparison.OrdinalIgnoreCase));
+            string str = "Turkish I \u0131s TROUBL\u0130NG!";
```
Comment (tarekgh) [inline src/System.Runtime/tests/System/StringTests.cs]: >string str = "Turkish I \u0131s TROUBL\u0130NG!"; [](start = 12, length = 49)

nit: maybe moving this inside the remote invoke be better?

File: src/System.Runtime/tests/System/StringTests.cs [hunk 6]
```diff
@@ -1293,147 +1293,157 @@ public static void IndexOf_AllSubstrings(string s, string value, int startIndex,
...
         }
 
         [Fact]
         public static void IndexOf_HungarianDoubleCompression()
         {
-            string source = "dzsdzs";
-            string target = "ddzs";
-            Helpers.PerformActionWithCulture(new CultureInfo("hu-HU"), () =>
+            string src = "dzsdzs";
+            string trgt = "ddzs";
```
Comment (tarekgh) [inline src/System.Runtime/tests/System/StringTests.cs]: nit: same as previous comment

File: src/System.Runtime/tests/System/StringTests.cs (hunk 5 above)
Comment (sepidehMS) [inline src/System.Runtime/tests/System/StringTests.cs]: Will do.

File: src/System.Runtime/tests/System/StringTests.cs [hunk 7]
```diff
@@ -2248,12 +2263,31 @@ private static IEnumerable<object[]> ToLower_Culture_TestData()
...
+                        CultureInfo.CurrentCulture = (CultureInfo)testdata[2];
+                        ToLower_Culture((string)testdata[0], (string)testdata[1]);
+                    }
+                    finally
+                    {
+                        CultureInfo.CurrentCulture = originalCulture;
+                    }
+                }
+                return SuccessExitCode;
+            }).Dispose();
```
Comment (tarekgh) [inline src/System.Runtime/tests/System/StringTests.cs]: you still need to fix this as you did for other cases

Comment (sepidehMS) [inline src/System.Runtime/tests/System/StringTests.cs]: In this case, CurrentCulture is an input coming from test data and changes for each case.

**Comment by danmosemsft** [comment]
@hughbe fyi as you were mentioning this somewhere.

File: src/System.Runtime/tests/System/StringTests.cs (hunk 7 above)
Comment (sepidehMS) [inline src/System.Runtime/tests/System/StringTests.cs]: I can move it to be done in the test and remove reverting to original culture, though.

Comment (tarekgh) [inline src/System.Runtime/tests/System/StringTests.cs]: ok, then you don't need the finally block. right?

---
In reply to: [123817517](https://github.com/dotnet/corefx/pull/21471#discussion_r123817517) [](ancestors = 123817517)

**Comment by tarekgh** [comment]
LGTM. Thanks @sepidehMS

**Comment by danmosemsft** [comment]
@tarekgh @hughbe it sounds like we need a pass to move every instances of setting CurrentCulture and CurrentUICulture to RemoteInvoke?

**Pull request merged**